src/tests/posix/pthread/pthread_rwlock_test.c

```diff
@@ -0,0 +1,93 @@
+/**
+ * @file
+ *
+ * @date Mar 02, 2020
+ * @author: Puranjay Mohan
+ */
+
+#include <pthread.h>
+#include <poll.h>
+#include <embox/test.h>
+#include <stdlib.h>
+#define ACCESS_ONCE(x) (*(volatile typeof(x) *) &(x))
+
+EMBOX_TEST_SUITE("posix/pthread_rwlock_api");
+
+struct data {
+	pthread_rwlock_t lock_rw;
+	int x;
+	int reader_x;
+};
+
+TEST_CASE ("Initialize rwlock with pthread_rwlock_init()") {
+
+	pthread_rwlock_t rw;
+	test_assert_zero(pthread_rwlock_init(&rw, NULL));
+	test_assert_zero(pthread_rwlock_rdlock(&rw));
+	//function pthread_rwlock_trywrlock() is not implemented hence returns ENOSYS
+	test_assert_equal(pthread_rwlock_trywrlock(&rw), -ENOSYS);
+	test_assert_zero(pthread_rwlock_unlock(&rw));
+	test_assert_zero(pthread_rwlock_wrlock(&rw));
+	//function pthread_rwlock_tryrdlock() is not implemented hence returns ENOSYS
+	test_assert_equal(pthread_rwlock_tryrdlock(&rw), -ENOSYS);
+	test_assert_zero(pthread_rwlock_unlock(&rw));
+	test_assert_zero(pthread_rwlock_destroy(&rw));
+}
+
+void *reader_thread(void *arg) {
+
+	int i;
+	int newx, oldx;
+	newx = oldx = -1;
+	struct data *lockd = (struct data *)arg;
+	pthread_rwlock_t *p = (pthread_rwlock_t *) (&(lockd->lock_rw));
+
+	for (i = 0; i < 100; i++) {
+		test_assert_zero(pthread_rwlock_rdlock(p));
+		newx = ACCESS_ONCE(lockd->x);
+		test_assert_zero(pthread_rwlock_unlock(p));
+		poll(NULL, 0, 1);
+		if (newx != oldx) {
+			lockd->reader_x = newx;
+		}
+		oldx = newx;
+	}
+
+	return NULL;
+}
+
+void *writer_thread(void *arg) {
+
+	int i;
+	struct data *lockd = (struct data *) arg;
+	pthread_rwlock_t *p = (pthread_rwlock_t *) (&(lockd->lock_rw));
+
+	for (i = 0; i < 3; i++) {
+		test_assert_zero(pthread_rwlock_wrlock(p));
+		ACCESS_ONCE(lockd->x)++;
+		test_assert_zero(pthread_rwlock_unlock(p));
+		poll(NULL, 0, 5);
+	}
+
+	return NULL;
+
+}
+
+TEST_CASE("Testing Reader Writer Lock") {
+
+	pthread_t tid1, tid2;
+	void *vp;
+	struct data *lock_data = (struct data *) malloc(sizeof(struct data));
```

Comment:
I think it's better to use `static struct data lock_data`
But your version is also correctly

Reply:
I think it works without the static, so keeping it like that.